DC-572: add ability to share links to filtered views in the data catalog

pshapiro4broad opened the pull request:
Add support for bookmarking and link sharing of a filtered catalog view. The selected tag state is stored in the URL as a user clicks in the UI. While doing this, I noticed that the selected tag state stores `section` and `label` but doesn't seem to use these fields, so I removed them.

Instead of adding a new test, I added a test case to the existing integration test for this page.

An example of the URL generated by this code:

```
https://app.terra.bio/#library/datasets?selectedTags%5BConsortium%5D%5B0%5D=radx-rad
```

pshapiro4broad committed:
code works, no test yet, needs better error handling

pshapiro4broad committed:
Merge branch 'dev' into ps-dc-572-filter-state-in-url

pshapiro4broad committed:
Merge branch 'dev' into ps-dc-572-filter-state-in-url

pshapiro4broad committed:
- fix bug where `clear` didn't clear the selected filter tags
- fix bug where clearing filter tags didn't result in a URL without any query params
- refactor common code for changing the filter tags in the URL

pshapiro4broad committed:
Merge remote-tracking branch 'origin/ps-dc-572-filter-state-in-url' into ps-dc-572-filter-state-in-url

pshapiro4broad committed:
add assert for URL rewrite to existing integration test

pshapiro4broad committed:
Merge branch 'dev' into ps-dc-572-filter-state-in-url

s-rubenstein commented on src/pages/library/common.js:
```diff
@@ -503,12 +503,11 @@ export const SearchAndFilterComponent = ({
-            })
+            const newTags = _.xorBy('lowerTag', [{ lowerTag }], selectedTags[section])
+            const newSelectedTags = newTags.length > 0 ? _.set(section, newTags, selectedTags) : _.omit(section, selectedTags)
+            navigateToFilterAndSelection(searchFilter, newSelectedTags)
```
Its a little weird to need to pass the searchFilter in here (and the existing tags on above.) However, on looking at it, I think this is still preferable to two methods with one arg that call a common helper for the navigation event.

pshapiro4broad replied:
~Is there a way to provide named parameters with defaults in javascript? I came across this post https://simonsmith.io/destructuring-objects-as-function-parameters-in-es6#allowing-the-configuration-object-to-be-optional but couldn't get it to work.~

Actually this does work fine (I had some other uncommitted changes that were breaking things). So now I have:
```
  const navigateToFilterAndSelection = ({ filter = searchFilter, tags = selectedTags } = {}) => {
```

and this call becomes:
```
            navigateToFilterAndSelection({ tags: newSelectedTags })
```

Is this kind of form used in terra-ui?

kyuksel replied:
yes

s-rubenstein commented on src/pages/library/common.js:
```diff
@@ -519,7 +518,7 @@ export const SearchAndFilterComponent = ({
       div({ style: { marginLeft: '1rem', minWidth: 0, width: '100%', height: '100%' } }, [
         _.isEmpty(filteredData.data) ? div({ style: { margin: 'auto', textAlign: 'center' } }, ['No Results Found']) :
-          children({ filteredList: filteredData.data, sections, selectedTags, setSelectedTags })
+          children({ filteredList: filteredData.data, sections })
```
Why don't you need to pass the selected tags down?

pshapiro4broad replied:
Honestly I don't know.. It looks like `children` is an argument to `SearchAndFilterComponent` (like `idField` or `descField`) but I don't see anyone passing it in, so I don't see how it's actually used.

pshapiro4broad replied:
If this could be the cause of a bug, what should I do to test for that bug? The UI seems to work fine with this change when I use it.

s-rubenstein replied:
I think if it looks like the UI looks fine, that is fine for now. Eventually we would have unit tests which would tell us, but I think unfortunately we don't have those yet, so we have to make do somewhat with manual testing.

s-rubenstein reviewed (approved):
LGTM, pending one question.

pshapiro4broad committed:
- use object type to implement default arguments

- only store the lower tag in the select tags state (and in the URL)

pshapiro4broad committed:
Merge remote-tracking branch 'origin/ps-dc-572-filter-state-in-url' into ps-dc-572-filter-state-in-url

pshapiro4broad committed:
- fix behavior of Filter By modal dialog to work with new way of storing filter selection

- fix issue with Filter By modal where Cancel would Apply instead of discarding changes

pshapiro4broad committed:
Merge branch 'dev' into ps-dc-572-filter-state-in-url

# Conflicts:
#	src/pages/library/common.js

pshapiro4broad commented on src/pages/library/common.js:
```diff
@@ -514,12 +511,11 @@ export const SearchAndFilterComponent = ({
-              return newTags.length > 0 ? _.set(section, newTags, oldSelectedTags) : _.omit(section, oldSelectedTags)
-            })
+          onTagFilter: ({ section, tags }) => {
+            _.forEach(tag => Ajax().Metrics.captureEvent(`${Events.catalogFilter}:sidebar`, { tag }), tags)
```
I thought about changing this event. Right now the `section` for the `tag` isn't sent, so the event doesn't know which section a tag might belong to. Also a user can filter multiple tags at once, using the `Filter by` dialog. I kept the old behavior which each filter is sent as a separate event, but this could be done with a single event too.

pshapiro4broad committed:
Restore custom OK button message for See more / Filter By dialog.
